src/hotspot/share/gc/shared/scavengableNMethods.hpp

```diff
@@ -46,8 +46,10 @@ class ScavengableNMethods : public AllStatic {
   static void unregister_nmethod(nmethod* nm);
   static void verify_nmethod(nmethod* nm);
 
-  // Remove nmethods that no longer have scavengable oops.
-  static void prune_nmethods();
+  // Remove nmethods that no longer have oops into young gen.
+  static void prune_nmethods_not_into_young();
+  // Remvoe unlinked (dead) nmethods.
+  static void prune_unlinked_nmethods();
```

Comment:
Can they be merged into one API, pruning nmethods that are either unlinked or doesn't point into young-gen?

Reply:
I left this out because at the time we unlink nmethods we have not moved the objects yet, only marked them. If we moved the class unloading to where `prune_nmethods_not_into_young` is, it might be possible. Going to file a CR to investigate this.

Reply:
Fwiw, I did not move the class unloading because in one of the last changes you seemed fairly insistent on keeping ordering and location of class unloading consistent across stw collectors too.

Reply:
> at the time we unlink nmethods we have not moved the objects yet

I was thinking of calling both at mark-end, but I overlooked that objs have not been relocated yet... The current impl looks good.